lib/puppet_x/puppetlabs/powershell_manager.rb

```diff
@@ -0,0 +1,169 @@
+require 'securerandom'
+require 'open3'
+require 'ffi'
+
+module PuppetX
+  module Dsc
+    class PowerShellManager
+      extend Puppet::Util::Windows::String
+      extend FFI::Library
+
+      @@instances = {}
+
+      def self.instance(cmd)
+        @@instances[:cmd] ||= PowerShellManager.new(cmd)
+      end
+
+      def initialize(cmd)
+        @output_ready_event_name =  "Global\\#{SecureRandom.uuid}"
+        @output_ready_event = self.class.create_event(@output_ready_event_name)
+
+        @stdin, @stdout, @ps_process = Open3.popen2(cmd)
+        @stdin.sync = true
+        @stdout.sync = true
+
+        Puppet.debug "#{Time.now} #{cmd} is running as pid: #{@ps_process[:pid]}"
+
+        at_exit { exit }
+      end
+
+      def execute(powershell_code)
+        # always need a trailing newline to ensure PowerShell parses code
+        out = exec_read_result(<<-CODE
+        $event = [Threading.EventWaitHandle]::OpenExisting("#{@output_ready_event_name}")
+
+        $Error.Clear()
+        $LASTEXITCODE = 0
+
+        #{powershell_code}
+
+        [Void]$event.Set()
```

Comment:
We already wrap `Invoke-DscResource` in a try/catch in these lines https://github.com/puppetlabs/puppetlabs-dsc/blob/master/lib/puppet/provider/templates/invoke_dsc_resource.ps1.erb#L62-L67, so a try/catch block works.

This however is after parameter assignment and after we create a credential, both operations which may throw exceptions.

I'm generally hesitant on adding general catch alls, but it may be worth removing the try/catch from inside the invoke script and moving it out to here, so we can use finally to ensure the event is set.